### PR TITLE
Adding P1 automation support for the Deferred Fog component

### PR DESCRIPTION
**Pull request by SWMasterson**
Adding P1 support to the Deferred Fog component to test:

- Toggle states of all switches (enable/disable).
- Min/Max values of all applicable parameters.
- All vector parameters.

Ran automation in editor and via TestSuite_Main.py:

Report:
[SUCCESS] Success: Deferred Fog Entity successfully created
[SUCCESS] Success: Entity has a Deferred Fog component
[SUCCESS] Success: Deferred Fog component successfully removed
[SUCCESS] Success: UNDO Deferred Fog component removal success
[SUCCESS] Success: Deferred Fog component disabled
[SUCCESS] Success: Entity has a PostFX Layer component
[SUCCESS] Success: Deferred Fog component enabled
[SUCCESS] Success: Enable Deferred Fog parameter enabled
[SUCCESS] Success: Enable Deferred Fog parameter disabled
[SUCCESS] Success: Enable Turbulence Properties parameter enabled
[SUCCESS] Success: Enable Turbulence Properties parameter disabled
[SUCCESS] Success: Enable Fog Layer parameter enabled
[SUCCESS] Success: Enable Fog Layer parameter disabled
[SUCCESS] Success: Fog Color parameter updated
[SUCCESS] Success: Fog Start Distance set to minimum value
[SUCCESS] Success: Fog Start Distance set to maximum value
[SUCCESS] Success: Fog End Distance set to maximum value
[SUCCESS] Success: Fog End Distance set to minimum value
[SUCCESS] Success: Fog Bottom Height set to maximum value
[SUCCESS] Success: Fog Bottom Height set to minimum value
[SUCCESS] Success: Fog Max Height set to minimum value
[SUCCESS] Success: Fog Max Height set to maximum value
[SUCCESS] Success: Noise Texture First Octave Scale updated
[SUCCESS] Success: Noise Texture First Octave Velocity updated
[SUCCESS] Success: Noise Texture Second Octave Scale updated
[SUCCESS] Success: Noise Texture Second Octave Velocity updated
[SUCCESS] Success: Octaves Blend Factor set to minimum value
[SUCCESS] Success: Octaves Blend Factor set to maximum value
[SUCCESS] Success: Entered game mode
[SUCCESS] Success: Exited game mode
[SUCCESS] Success: Entity is hidden
[SUCCESS] Success: Entity is visible
[SUCCESS] Success: Entity deleted
[SUCCESS] Success: UNDO deletion success
[SUCCESS] Success: REDO deletion success
Test result:  SUCCESS

.\python\python.cmd -m pytest --build-directory build_development_new\bin\profile  .\AutomatedTesting\Gem\PythonTests\Atom\TestSuite_Main.py
==== test session starts ====
platform win32 -- Python 3.7.12, pytest-5.3.2, py-1.11.0, pluggy-0.13.1
rootdir: F:\git\o3de, inifile: pytest.ini
plugins: mock-2.0.0, timeout-1.3.4, ly-test-tools-1.0.0
collected 29 items

AutomatedTesting\Gem\PythonTests\Atom\TestSuite_Main.py .............................. [100%]

=== warnings summary ===
tools\lytesttools\ly_test_tools\o3de\editor_test.py:333
  f:\git\o3de\tools\lytesttools\ly_test_tools\o3de\editor_test.py:333: PytestCollectionWarning: cannot collect test class 'TestData' because it has a __init__ constructor (from: AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main.py::TestAutomation)
    class TestData:

AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main.py::TestMaterialEditorBasicTests::test_MaterialEditorBasicTests[windows-MaterialEditor-windows_generic-AutomatedTesting]
  f:\git\o3de\tools\lytesttools\ly_test_tools\_internal\pytest_plugin\test_tools_fixtures.py:298: DeprecationWarning: This method is deprecated and will be removed on 3/31/22. Please use another helper method instead.
    return ly_test_tools.launchers.launcher_helper.create_generic_launcher(workspace, launcher_platform, exe_file_name)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=== 30 passed, 2 warnings in 190.84s (0:03:10) ===

Signed-off-by: Sean Masterson <semaster@amazon.com>

### CODE REVIEW

**Review by gadams3 — APPROVED** [review]

File: AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py [hunk 1]
```diff
@@ -75,19 +135,33 @@ def AtomEditorComponents_DeferredFog_AddedToEntity():
     5) Verify Deferred Fog component not enabled.
```
Comment (jromnoa) [inline AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py:135]: The review doesn't let me target it, but lines 73-74 for steps 3 and 4 above this code do not match the commented code in the test script.
From lines 73-74:
```
3) UNDO the entity creation and component addition.
4) REDO the entity creation and component addition.
```

From line 188 and line 194:
```
3. Remove the Deferred Fog component.
4. Undo Bloom component removal.
```

**Review by jromnoa — CHANGES_REQUESTED** [review]

File: AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py (hunk 1 above)
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py:135]: Good catch, thank you. I didn't notice that I forgot to update the docstring on my readthrough.
Fixed.

File: AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py [hunk 2]
```diff
@@ -111,78 +185,236 @@ def AtomEditorComponents_DeferredFog_AddedToEntity():
...
+        # Update the Fog Max Height parameter to its maximum value.
+        deferred_fog_component.set_component_property_value(
+            AtomComponentProperties.deferred_fog('Fog Max Height'), 5000.0)
+        Report.result(Tests.fog_max_height_max,
+                      deferred_fog_component.get_component_property_value(
+                          AtomComponentProperties.deferred_fog('Fog Max Height')) == 5000.0)
+        general.idle_wait_frames(1)
+
+        # 16. Edit the Noise Texture parameter.
+        # This field cannot currently be edited. It will be fixed in the near future.
```
[the remark this answers is not in the document]
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py:341]: Noted that it'll be fixed in a future sprint.

File: AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py [hunk 3]
```diff
@@ -205,13 +205,43 @@ def deferred_fog(property: str = 'name') -> str:
...
+          - 'Noise Texture First Octave Scale' Scale of the first noise octave (INF, INF).
+            Higher values indicates higher frequency. Values set using azlmbr.math.Vector2 tuple.
+          - 'Noise Texture First Octave Velocity' Velocity of the first noise octave UV coordinates (INF, INF).
+            Values set using azlmbr.math.Vector2 tuple.
+          - 'Noise Texture Second Octave Scale' Scale of the second noise octave (INF, INF).
+            Higher values indicates higher frequency. Values set using azlmbr.math.Vector2 tuple.
+          - 'Noise Texture Second Octave Velocity' Velocity of the second noise octave UV coordinates (INF, INF).
+            Values set using azlmbr.math.Vector2 tuple.
+          - 'Octaves Blend Factor' Blend factor between the noise octaves (0.0, 1.0).
+          - 'Enable Turbulence Properties' Enables Turbulence Properties.
```
[the remark this answers is not in the document]
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py:223]: Updated to list this as (bool)

File: AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py [hunk 4]
```diff
@@ -205,13 +205,43 @@ def deferred_fog(property: str = 'name') -> str:
...
+          - 'Enable Fog Layer' Enables the fog layer.
```
[the remark this answers is not in the document]
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py:224]: Updated to list this as (bool)

File: AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py [hunk 5]
```diff
@@ -205,13 +205,43 @@ def deferred_fog(property: str = 'name') -> str:
           - 'requires' a list of component names as strings required by this component.
             Use editor_entity_utils EditorEntity.add_components(list) to add this list of requirements.\n
           - 'Enable Deferred Fog' Toggle active state of the component True/False
+          - 'Fog Color' Sets the fog color. RGB value set using azlmbr.math.Vector3 tuple.
+          - 'Fog Start Distance' Distance from the viewer where the fog starts (0.0, 5000.0).
+          - 'Fog End Distance' Distance from the viewer where fog masks the background scene (0.0, 5000.0).
+          - 'Fog Bottom Height' Height at which the fog layer starts (-5000.0, 5000.0).
+          - 'Fog Max Height' Height of the fog layer top (-5000.0, 5000.0).
+          - 'Noise Texture' The noise texture used for creating the fog turbulence.
```
[the remark this answers is not in the document]
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py:213]: Listed it as Asset.id (using other similar components as an example).
Also noted here that it is not yet implemented for editing and will be fixed in a future sprint.

File: AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py [hunk 6]
```diff
@@ -111,78 +185,236 @@ def AtomEditorComponents_DeferredFog_AddedToEntity():
...
+        # Update the Fog End Distance parameter to its maximum value.
+        deferred_fog_component.set_component_property_value(
+            AtomComponentProperties.deferred_fog('Fog End Distance'), 5000.0)
+        Report.result(Tests.fog_end_max,
+                      deferred_fog_component.get_component_property_value(
+                          AtomComponentProperties.deferred_fog('Fog End Distance')) == 5000.0)
+
+        # Update the Fog End Distance parameter to its minimum value.
+        deferred_fog_component.set_component_property_value(
+            AtomComponentProperties.deferred_fog('Fog End Distance'), 0.0)
```
Comment (smurly) [inline AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py:304]: is there anything that prevents you from setting the end distance to zero while the start distance is 5000? it seems like it should but these getter/setters seem to be fine with taking anything in individual value range.

File: AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py (hunk 4 above)
Comment (smurly) [inline AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py]: nit: I'm not sure how 'Enable Fog Layer' is different than 'Enable Deferred Fog'

**Review by smurly — APPROVED** [review]

File: AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py (hunk 4 above)
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/atom_utils/atom_constants.py]: Enable deferred fog is the toggle for the active state of the component. When it's off, nothing else on the component registers in the editor.

The Enable Fog Layer is an actual layer of fog applied to the scene as opposed to just general fogginess throughout the entire scene. That's the best description I can give it anyway.

File: AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py (hunk 6 above)
Comment (SWMasterson) [reply AutomatedTesting/Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_DeferredFogAdded.py:304]: Nope, you're free to set them to whatever you want. There's nothing that stops you from setting start distance higher than end distance. Currently its up to the user to know what to do.

**Review by jromnoa — APPROVED** [review]

**Pull request merged**